### PR TITLE
Application: Decentralized eclipse chat 

### PR DESCRIPTION
**Pull request by herou**
#### Project Abstract

We aim to build a decentralized communication system that will help businesses/clients to communicate.
We will use substrate node template to build our pallet.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] A BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (aUSD/USDT) address for the payment of the milestones is provided in the application.
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [x] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@elio.prifti:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application. Let me share this with the people that evaluated your previous deliveries. But I also have some initial questions/comments:

- What are your long-term goals with this project, and are you going to maintain it? For example, you don’t seem to use or maintain the [escrow pallet](https://github.com/herou/EscrowPallet/tree/eljo-prifti/escrow), or am I missing something? 
- In general, the application seems relatively expensive since you are asking for 30k for a single pallet.

**Comment by herou** [comment]
> Thanks for the application. Let me share this with the people that evaluated your previous deliveries. But I also have some initial questions/comments:
> 
> * What are your long-term goals with this project, and are you going to maintain it? For example, you don’t seem to use or maintain the [escrow pallet](https://github.com/herou/EscrowPallet/tree/eljo-prifti/escrow), or am I missing something?
> * In general, the application seems relatively expensive since you are asking for 30k for a single pallet.

Thank you for the quick reply @Noc2 
We have a very detailed plan for this project. In a few words, we would like to build an android application for this communication system and maintain it for a long time.
The escrow pallet was just an example for the community (my very first substrate project) and we thought not many people will use it compared to the eclipseChat we aim to build.
Regarding the amount of money, even though is a single pallet, it contains 9 methods. Just to mention, the Escrow pallet contains 3 methods. The amount of work is 3 times more than the Escrow pallet we have built.
We would be very pleased if web3 would give us the opportunity to build this product. 
Thanks a lot!

**Review by 0xCaso — COMMENTED** [review]
Hi @herou, thanks for the application. I evaluated your previous grant "decentralized invoice". That said, and after I read the application, I can say:
- You can remove the **Article** deliverable from the milestones and just put it into the last one, as you won't update it every time (at least, this is what I remember from the previous grant).
- In my opinion, as you'll add three methods to the pallet for each milestone, the **Development Roadmap** needs to be modified: both **duration** and **costs** (for each milestone) seem a bit excessive. As it stands, it would be reasonable that the first milestone could take a bit of time, as you need to set the project up. But after that, the following milestones should be straightforward, and definitely must be way shorter than the first one (as the project is already set up). 
For example, in the decentralized invoice project, you basically had everything ready from the first milestone (see [here](https://github.com/gertt/Invoice/blob/524e528ada7db2cbb187cada070c447899e8bee2/pallets/invoice/src/lib.rs)), and you just needed to edit something for each delivery. So, I think it could be a nice idea to put everything in a single milestone, and adjust both **duration** and **costs**.
- As you say, you're willing to maintain the project, so it would be nice to restructure the application including a milestone for the **frontend**, if you're interested. This way, the application would have two milestones: one for the **backend** (the pallet) and one for the **frontend**. This way, I think the grant could be a **Level 2**, as it would be more structured.
- Finally, are you aware of already existent projects in our ecosystem that bring messaging on-chain, such as [Uke](https://github.com/CrackTheCode016/Grants-Program/blob/9dad8190a53d717a10c2bccb4b3c39269474cbf8/applications/uke-protocol.md)? How is your project different from, for example, it? (For other examples, go to the [Open Source Stack](https://wiki.polkadot.network/docs/build-open-source#chains-and-pallets) in the Polkadot Wiki -> Chains and Pallets -> Messaging).

File: applications/decentralized_eclipse_chat_2.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+We would like to use the Substrate template node
+- [Substrate node template](https://github.com/substrate-developer-hub/substrate-node-template)
+
+## Development Roadmap
+
+### Overview
+
+- **Total Estimated Duration:** 3 months
+- Full-Time Equivalent (FTE): 2 FTE
+- **Costs:** $30,000 USDT
```
Comment (0xCaso) [inline applications/decentralized_eclipse_chat_2.md:146]: Also, please could you add "Total" here to make our integrations work?
```suggestion
- **Total Costs:** $30,000 USDT
```

**Comment by herou** [comment]
> Hi @herou, thanks for the application. I evaluated your previous grant "decentralized invoice". That said, and after I read the application, I can say:
> 
> * You can remove the **Article** deliverable from the milestones and just put it into the last one, as you won't update it every time (at least, this is what I remember from the previous grant).
> * In my opinion, as you'll add three methods to the pallet for each milestone, the **Development Roadmap** needs to be modified: both **duration** and **costs** (for each milestone) seem a bit excessive. As it stands, it would be reasonable that the first milestone could take a bit of time, as you need to set the project up. But after that, the following milestones should be straightforward, and definitely must be way shorter than the first one (as the project is already set up).
>   For example, in the decentralized invoice project, you basically had everything ready from the first milestone (see [here](https://github.com/gertt/Invoice/blob/524e528ada7db2cbb187cada070c447899e8bee2/pallets/invoice/src/lib.rs)), and you just needed to edit something for each delivery. So, I think it could be a nice idea to put everything in a single milestone, and adjust both **duration** and **costs**.
> * As you say, you're willing to maintain the project, so it would be nice to restructure the application including a milestone for the **frontend**, if you're interested. This way, the application would have two milestones: one for the **backend** (the pallet) and one for the **frontend**. This way, I think the grant could be a **Level 2**, as it would be more structured.
> * Finally, are you aware of already existent projects in our ecosystem that bring messaging on-chain, such as [Uke](https://github.com/CrackTheCode016/Grants-Program/blob/9dad8190a53d717a10c2bccb4b3c39269474cbf8/applications/uke-protocol.md)? How is your project different from, for example, it? (For other examples, go to the [Open Source Stack](https://wiki.polkadot.network/docs/build-open-source#chains-and-pallets) in the Polkadot Wiki -> Chains and Pallets -> Messaging).

Hi @0xCaso 
Thank you for your reply.

- If you take a closer look, the article has been included in all milestones to the previous grant application (decentralized invoice) and I think it should be the same for this one as well.
- We are going to use the Substrate node template, there is no project to set up (it takes 10 minutes to download the Substrate node template and to add a pallet on top of it), which means the first milestone takes as much time as the other milestones.  Please, check my previous grant proposals, I have done the same.
- Building our backend on Substrate and the android application within 3 months is too much because we are just 2 devs. Our focus would be the blockchain first, writing all the functionalities, unit tests, and benchmarks. We can give details about the android app as well if you want, but we CAN NOT include it in one of the milestones. 
- There are a lot of messaging apps built in web2 as well like Telegram, Whatsapp, Discord, and Signal and all of them have their own clients, even though they provide the same service. What makes us different from our competitors, we would be a more gaming-oriented application. For the moment we need the basic functionalities. Future development will come.

I hope it is clear to you. If something is still not clear, please let me know. Thanks again :)

**Review by keeganquigley — CHANGES_REQUESTED** [review]
Hi @herou thanks for the application. Here are my initial questions:

- Can you please include the currency for payment (DAI, USDC, etc.) Please note that our payment options have [changed recently](https://github.com/w3f/Grants-Program#pencil-process).
- I agree with @0xCaso that in your previous grant deliveries, you did significantly more work on the first milestone than the subsequent ones, which were more straightforward optimizations. If you don't want to work on the front-end for this grant, I would also recommend combining the work into one or two milestones.
- What encryption method(s) do you plan on using? Have you researched possible attack vectors?
- Can you please add more technical details about how you are planning to integrate the pallets with SMS messages? Would you for example be communicating with an API service to send the messages, or how does this work?
- I'm not seeing that you wrote three separate articles in previous grants, only [one](https://medium.com/@elioprifti/escrow-pallet-build-on-top-of-substrate-framework-6f941c28b6fd). So like @0xCaso mentioned, you can go ahead and remove this from the first two milestones and only include in the last one.
- It's also fair to say that with a level 2 grant comes more scrutinization, so additional committee members may also chime in and ask questions. Thanks!

**Comment by herou** [comment]
> Hi @herou thanks for the application. Here are my initial questions:
> 
> * Can you please include the currency for payment (DAI, USDC, etc.) Please note that our payment options have [changed recently](https://github.com/w3f/Grants-Program#pencil-process).
> * I agree with @0xCaso that in your previous grant deliveries, you did significantly more work on the first milestone than the subsequent ones, which were more straightforward optimizations. In fact you submitted two at the same time. If you don't want to work on the front-end for this grant, I would also recommend combining the work into one or two milestones.
> * What encryption method(s) do you plan on using? Have you researched possible attack vectors?
> * Can you please add more technical details about how you are planning to integrate the pallets with SMS messages? Would you for example be communicating with an API service to send the messages, or how does this work?
> * I'm not seeing that you wrote three separate articles in previous grants, only [one](https://medium.com/@elioprifti/escrow-pallet-build-on-top-of-substrate-framework-6f941c28b6fd). So like @0xCaso mentioned, you can go ahead and remove this from the first two milestones and only include in the last one.
> * It's also fair to say that with a level 2 grant comes more scrutinization, so additional committee members may also chime in and ask questions. Thanks!

- Done
- I submitted two milestones at the same time because the first milestone was delayed, so I decided to submit the first and the second milestone at the same time.
- I would like to keep my milestones as they are, It has been planned as a team and I can not change them.
- Https for communication with Node, Blake2-128 for storing data on-chain. Authentication of every user before communication with the Node

- Let me give an example of the whole system.
	1. A client opens the android app
	2. He gets authenticated using his wallet
	3. The address and his nickname(should be set) get stored in Blockchain. If both of them are not already stored (check the Nickname method that would be used)
	4. The client search for an address and send an invitation in order to open a communication between two addresses
	If the invitation/flag is accepted, a structure/channel would be created between these two clients, and all the messages
will be stored in this structure.
	6. Send/get function would be used in order to send and receive msg between these two addresses.
	7. Get all 3 channels will show up all the created channels for an address
	8. If one of the clients wants to delete the chat, then in storage all the data would be replaced with zeros or other meaningless characters. So the communication can not be read anymore.
	
WhatsApp Image 2023-02-01 at 19 51 47



- Done
- No problem, I am here to answer all the questions.

**Comment by keeganquigley** [comment]
thanks for the updates @herou I will go ahead and mark your application as ready for review and ping the committee, but if you could add these details into the application along with the mock-up that would be great.

**Comment by herou** [comment]
> thanks for the updates @herou I will go ahead and mark your application as ready for review and ping the committee, but if you could add these details into the application along with the mock-up that would be great.

Actually, we have not worked on the mock-up yet. Since we thought to develop it in the later stage, but if we want we can send some mock-up at the end of milestone one!

**Comment by keeganquigley** [comment]
thanks @herou I just meant if you could include the above information into the application itself.

**Comment by herou** [comment]
> thanks @herou I just meant if you could include the above information into the application itself.

There is a lot of information. I am not sure which one should be included :/

**Comment by keeganquigley** [comment]
Sorry, I was referring to the image you posted of the app architecture, as well as your summary of how the app will work. We ask that the technical specifics be included in the application itself, that way it can be easily found and referred to when searching our records. Thank you!

**Comment by 0xCaso** [comment]
Sorry if I reply only now @herou:
> * We are going to use the Substrate node template, there is no project to set up (it takes 10 minutes to download the Substrate node template and to add a pallet on top of it), which means the first milestone takes as much time as the other milestones.  Please, check my previous grant proposals, I have done the same.

Great, then also the first milestone's duration can be adjusted if the setup is straightforward.

> * Building our backend on Substrate and the android application within 3 months is too much because we are just 2 devs. Our focus would be the blockchain first, writing all the functionalities, unit tests, and benchmarks. We can give details about the android app as well if you want, but we CAN NOT include it in one of the milestones.

In two people, I think 2 months are more than enough to build the backend you're describing, but 3 months is too much time in my honest opinion. Optimistically speaking, building it in 1 month could be more than doable. Please, take a look into [Uke](https://github.com/w3f/Grants-Program/blob/master/applications/uke-protocol.md) application, where the grantee proposed (and built) a similar application (both backend and frontend, a month for each).

> * There are a lot of messaging apps built in web2 as well like Telegram, Whatsapp, Discord, and Signal and all of them have their own clients, even though they provide the same service. What makes us different from our competitors, we would be more mobile native oriented, we point to the quality, and the IOS app will follow. For the moment we need the basic functionalities. Future development will come.

Uke has provided an implementation for Android, iOS, and web. By the way yes, competition is important and well accepted if aims to fill other projects gaps, so let's see what the Committe members think about it.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the reply here. I agree with the other reviewers. Based on your previous deliveries and the current scope/price ratio of the applications, I won’t approve. Additionally, I’m personally currently not convinced that you have a “very detailed plan for this project” and that you will maintain it longer-term. The application itself doesn’t contain this information, as far as I can see. But maybe other reviewers have a different opinion.

**Comment by herou** [comment]
> Thanks for the reply here. I agree with the other reviewers. Based on your previous deliveries and the current scope/price ratio of the applications, I won’t approve. Additionally, I’m personally currently not convinced that you have a “very detailed plan for this project” and that you will maintain it longer-term. The application itself doesn’t contain this information, as far as I can see. But maybe other reviewers have a different opinion.

@Noc2 As I told you we would like to build something big and for long term. What other information do you need? We can work on that and update the application. Give us the positivity to build a great procuct. Please, reconsider it and let us know. I am sure we can make it work. Thank you!

**Comment by herou** [comment]
@Noc2 @0xCaso @keeganquigley  I am closing this PR

**Pull request closed without merging**